rust/src/io/object_store.rs

```diff
@@ -83,45 +85,42 @@ impl ObjectStore {
             return Ok(Self::memory());
         };
 
-        let parsed = match Url::parse(uri) {
-            Ok(u) => u,
-            Err(ParseError::RelativeUrlWithoutBase) => {
-                let str_path = tilde(uri).to_string();
-                let path = std::path::Path::new(&str_path);
-                return Ok(Self {
-                    inner: Arc::new(LocalFileSystem::new()),
-                    scheme: String::from("file"),
-                    base_path: Path::from(path.absolutize()?.to_str().unwrap()),
-                    prefetch_size: 4 * 1024,
-                });
-            }
-            Err(e) => {
-                eprintln!("Parse err: {e}");
-                return Err(Error::IO(format!("URI parse error: {e}")));
-            }
-        };
-
-        let scheme: String;
-        let object_store: Arc<dyn OSObjectStore> = match parsed.scheme() {
-            "s3" => {
-                scheme = "s3".to_string();
-                build_s3_object_store(uri).await?
-            }
-            "file" => {
-                scheme = "flle".to_string();
-                Arc::new(LocalFileSystem::new())
-            }
-            &_ => todo!(),
-        };
+        // Try to parse the provided string as a Url, if that fails treat it as local FS
+        future::ready(Url::parse(uri))
+            .map_err(Error::from)
+            .and_then(|url| Self::new_from_url(url))
+            .or_else(|_| future::ready(Self::new_from_path(uri)))
+            .await
+    }
 
+    pub fn new_from_path(str_path: &str) -> Result<Self> {
+        let expanded = tilde(str_path).to_string();
+        let absolute_path = StdPath::new(expanded.as_str()).absolutize()?;
+        let absolute_path_str = absolute_path
+            .to_str()
+            .ok_or(Error::IO(format!("can't convert path {}", str_path)))?;
+        let url = Url::from_file_path(absolute_path_str).unwrap();
         Ok(Self {
-            inner: object_store,
-            scheme,
-            base_path: Path::from(parsed.path()),
+            inner: Arc::new(LocalFileSystem::new()),
+            scheme: String::from("flle"),
+            base_path: Path::from(url.path()),
             prefetch_size: 64 * 1024,
         })
     }
 
+    pub async fn new_from_url(url: Url) -> Result<Self> {
```

Comment:
Can we consolidate `new_from_uri` and `new_from_path`?

It introduces more APIs for users to understand / learn.